### PR TITLE
CU-2d9gz9p - Refine to adapt testnet.

### PR DESCRIPTION
**Pull request by fred-yu-2013**

### CODE REVIEW

File: tests/src/res/custom.json [hunk 1]
```diff
@@ -19,6 +19,7 @@
   },
 
   "node": {
+    "provider": "http://localhost:9001",
```
Comment (carlduranleau) [inline tests/src/res/custom.json:22]: This property was removed few weeks ago because it's a duplicate of targetHost. So if 'provider' is a better naming, we should simply rename it everywhere to keep only one property.

File: tests/src/did/didentity.ts [hunk 2]
```diff
@@ -22,110 +19,51 @@ export class DIDEntity {
...
+				console.info(`${this.name}: identity synchronized result: ${synced}`);
+				this.did = rootIdentity.getDid(0);
+			} else {
+				const doc: DIDDocument = await rootIdentity.newDid(this.storepass);
+				this.did = doc.getSubject();
+				console.info(`[${this.name}] My new DID created: ${this.did.toString()}`);
+			}
+		}
+		if (!this.did) {
+			console.error("Can not get the did from the local store.");
```
Comment (carlduranleau) [inline tests/src/did/didentity.ts:53]: Replace console.log with a logger.

File: tests/src/payment/paymentservice.test.ts [hunk 3]
```diff
@@ -28,36 +27,36 @@ describe("test payment service", () => {
         console.log("Payment Service version: " + paymentVersion);
```
Comment (carlduranleau) [inline tests/src/payment/paymentservice.test.ts:27]: Replace console.log with a logger.

File: tests/src/provider/providerservice.test.ts [hunk 4]
```diff
@@ -36,9 +35,13 @@ describe("test provider service", () => {
...
-        let backups: BackupDetail[] = await providerService.getBackups();
-        expect(backups).not.toBeNull();
-        expect(backups).not.toEqual([]);
+        try {
+            let backups: BackupDetail[] = await providerService.getBackups();
+            expect(backups).not.toBeNull();
+            expect(backups).not.toEqual([]);
+        } catch (e) {
+            expect(e).toBeInstanceOf(NotFoundException);
+        }
```
Comment (carlduranleau) [inline tests/src/provider/providerservice.test.ts:44]: Invalid test because we expect two different valid behaviors. Is the valid result an HTTP 404 error or an empty structure?

File: tests/src/subscribe/subscribe.test.ts [hunk 5]
```diff
@@ -32,7 +37,7 @@ describe("test vault subscribe function", () => {
         expect(version).not.toBeNull();
     });
 
-    test("testPlaceOrder", async() => {
+    test.skip("testPlaceOrder", async() => {
```
Comment (carlduranleau) [inline tests/src/subscribe/subscribe.test.ts:40]: With now working PaymentService tests, there's no need to skip this test since it's a duplicate of paymentservice.test.ts#testPaceOrder. Just like "testGetOrder".

File: src/http/httpclient.ts [hunk 6]
```diff
@@ -60,7 +60,8 @@ export class HttpClient {
     constructor(serviceContext: ServiceContext, withAuthorization: boolean, httpOptions: HttpOptions) {
         this.serviceContext = serviceContext;
         this.withAuthorization = withAuthorization;
-        this.httpOptions = this.validateOptions(httpOptions);
+        // this.httpOptions = this.validateOptions(httpOptions);
```
Comment (carlduranleau) [inline src/http/httpclient.ts:63]: If not useful, we can get rid ot comments.

**Review by carlduranleau — CHANGES_REQUESTED** [review]
Tests should run faster with a TestData singleton, but it had been made this way to prevent running tests on a possibly altered context. But if all tests are running fine this way, that can only be good news! 

I don't know if this PR is a WIP because some cleaning is needed and PaymentService tests are disabled, even if this PR is supposed to implement the Testnet configuration. If your PR was a WIP, please prefix the title with "WIP: " so no one would start a review on it.

File: src/http/httpclient.ts [hunk 7]
```diff
@@ -249,4 +250,19 @@ export class HttpClient {
...
+        return {
+            protocol: url.protocol,
+            host: url.host,
+            port: url.port ? url.port : (url.protocol == 'https:' ? '443' : '80'),
+            method: httpOptions.method ?? HttpClient.DEFAULT_METHOD,
+            path: url.pathname,
+            headers: httpOptions.headers ?? HttpClient.DEFAULT_HEADERS,
+            timeout: httpOptions.timeout ?? HttpClient.DEFAULT_TIMEOUT
+        };
+    }
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply src/http/httpclient.ts:267]: This update is for the default port setting is invalid and the old method is very complex and can be simplify.

[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply src/http/httpclient.ts:267]: This update is for the default port setting is invalid and the old method is very complex and can be simplified.

File: src/http/httpclient.ts [hunk 8]
```diff
@@ -249,4 +250,19 @@ export class HttpClient {
 
         return httpOptions;
     }
+
+    private getHttpOptionsByProviderAddress(httpOptions: HttpOptions): HttpOptions {
+        checkNotNull(httpOptions, "No HTTP configuration provided");
+        const providerAddress = this.serviceContext.getProviderAddress();
+        const url = new URL(providerAddress);
+        return {
+            protocol: url.protocol,
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply src/http/httpclient.ts:259]: The provider address already contains the protocol. I think we should use it.

File: src/http/httpclient.ts [hunk 9]
```diff
@@ -249,4 +250,19 @@ export class HttpClient {
...
+            host: url.host,
+            port: url.port ? url.port : (url.protocol == 'https:' ? '443' : '80'),
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply src/http/httpclient.ts:261]: But this will change other codes. So did a compatible change.

File: tests/src/config/testdata.ts [hunk 10]
```diff
@@ -37,21 +38,20 @@ export class TestData {
 		return `${prefix}_${Date.now().toString()}`;
 	}
 
-    public static async getInstance(testName: string, clientConfig: any, userDir?: string): Promise<TestData> {
-		Utils.checkNotNull(clientConfig, "Test configuration cannot be empty");
-		Utils.checkNotNull(clientConfig.node, "A valid test configuration is mandatory");
-		if (!userDir) {
-			userDir = TestData.USER_DIR;
-		}
+    public static async getInstance(testName: string): Promise<TestData> {
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply tests/src/config/testdata.ts:41]: After this change, we can run all tests with the same configuration. This is a requirement for releasing the new version of the hive node. And also every commit for hive js, we also need to make all test cases pass.

File: tests/src/res/custom.json (hunk 1 above)
Comment (fred-yu-2013) [reply tests/src/res/custom.json:22]: the `provider` address is for the hive node to keep the vault. The `targetHost` is for the hive node to keep the backup data.

File: tests/src/did/didentity.ts [hunk 11]
```diff
@@ -22,110 +19,51 @@ export class DIDEntity {
...
-		await this.identity.synchronize();
-
-		if (this.didString !== undefined){
-			this.identity.setDefaultDid(this.didString);
-			let defaultDid = this.identity.getDefaultDid();
-			DIDEntity.LOG.info("************************************* default DID: {}", defaultDid.toString());
-		}
-
+	public async initDid2(mnemonic: string, needResolve: boolean) {
+		const path = "data/didCache" + File.SEPARATOR + this.name;
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply tests/src/did/didentity.ts:25]: Yes, but we can change this later.

File: tests/src/did/didentity.ts [hunk 12]
```diff
@@ -22,110 +19,51 @@ export class DIDEntity {
...
-		
-		await this.identity.synchronize();
-
-		if (this.didString !== undefined){
-			this.identity.setDefaultDid(this.didString);
-			let defaultDid = this.identity.getDefaultDid();
-			DIDEntity.LOG.info("************************************* default DID: {}", defaultDid.toString());
-		}
-
+	public async initDid2(mnemonic: string, needResolve: boolean) {
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply tests/src/did/didentity.ts:24]: Yes

File: tests/src/payment/paymentservice.test.ts [hunk 13]
```diff
@@ -28,36 +27,36 @@ describe("test payment service", () => {
...
     });
 
-    test("testPlaceOrder", async () => {
+    test.skip("testPlaceOrder", async () => {
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply tests/src/payment/paymentservice.test.ts:30]: This is aligning with hive java. Because we need keep all test cases for one run.

File: tests/src/provider/providerservice.test.ts (hunk 4 above)
Comment (fred-yu-2013) [reply tests/src/provider/providerservice.test.ts:44]: If we split this test case with two. It is a little complex. We need to create two backup services for line #41 and remove all backup services for line #43. So making a test case like this is a better choice.

File: tests/src/res/developing.json [hunk 14]
```diff
@@ -1,23 +1,25 @@
 {
-  "resolverUrl": "https://api-testnet.elastos.io/newid",
+  "resolverUrl": "testnet",
 
   "application": {
     "name": "developApp",
     "mnemonic": "wash busy tell relief street maximum dumb fat legend end panic warrior",
+    "mnemonics2":"wash busy tell relief street maximum dumb fat legend end panic warrior",
```
[the remark this answers is not in the document]
Comment (fred-yu-2013) [reply tests/src/res/developing.json:7]: Because the TestData use `mnemonics2`, will fix this with TestData.

File: tests/src/subscribe/subscribe.test.ts (hunk 5 above)
Comment (fred-yu-2013) [reply tests/src/subscribe/subscribe.test.ts:40]: Align with hive java.

File: src/http/httpclient.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (carlduranleau) [reply src/http/httpclient.ts:232]: DEFAULT_PORT is used only in HttpClient as the default value for DEFAULT_OPTIONS. I can remove that later.
